pkg/network/server.go

```diff
@@ -439,6 +440,23 @@ func (s *Server) ConnectedPeers() []string {
 	return peers
 }
 
+// PeerHeights returns the block height and user agent of all connected peers.
+func (s *Server) PeerHeights() []result.PeerHeight {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	states := make([]result.PeerHeight, 0, len(s.peers))
+	for k := range s.peers {
+		states = append(states, result.PeerHeight{
+			Address:   k.PeerAddr().String(),
+			UserAgent: string(k.Version().UserAgent),
+			Height:    k.LastBlockIndex(),
```

Comment:
It's not very reliable. It's updated on:
 * initial handshake
 * ping/pong

So it can easily be outdated. For C# nodes it tends to be more up to date because of https://github.com/neo-project/neo/issues/1844, but we can't rely on that.

On the other hand, this makes the RPC safe, it just returns whatever the node knows without generating any P2P traffic. Don't know if it's useful this way.

Reply:
How often does ping/pong happen?

I can say it's at least still useful for *me*, for example [here](https://panel.fs.neo.org/getobject?cid=BCKoihvHSJM8X9JWY1ngqKXNcTofhe8TRUqyzRt9632e&oid=BJqE6VdZVheyyzbr9mXnQB8NUPHsdB3LdLt6vMJP8r92) you can see the response showing the C# nodes getting stuck on TestNet. But whether it's useful enough to be worth adding to NeoGo formally is another question.

Reply:
>How often does ping/pong happen?

It's a node setting, see the https://github.com/nspcc-dev/neo-go/blob/master/docs/node-configuration.md#p2p-configuration.

Reply:
> Don't know if it's useful this way.

Well, there's at least one person who finds it useful. To be honest, I also faced with the necessity to learn peers info, so I vote for adding this extension.

Reply:
C# nodes send pings for every block, Go nodes do this only if they're not progressing (no new blocks) for `PingInterval` time.

Reply:
PingInterval for mainnet/testnet NeoGo nodes is 30s, it's quite large interval, so we may end up with a really *stale* data. So the question is do we need these stale data, right?

Reply:
Go nodes _never_ send any pings at all if they're OK, it's not just about 30s.
https://github.com/nspcc-dev/neo-go/blob/cf4d4a2611209d8d085ccec6e1f1b8956006a577/pkg/network/server.go#L565

Reply:
Yes, I saw this code, but is it *that* never? I thought that the situation when RPC node is two blocks behind from the network (which is enough to send ping) is not a rare thing.

But if it is, then OK, this API is useless.

Reply:
You can check it by raising PingInterval to some insane value. But it's statistics anyway, can happen, can not happen.

Reply:
Hmm. Well, my use case for it is to act as a network health check:

1. What node versions are being run
2. Which ones are making progress

I don't really need to know if a NeoGo node is at the latest height. If it's a NeoGo node on the same version as me, I can already infer they are probably okay (config differences and operator error aside). Even if a very "ok" NeoGo node looks stuck because it's not pinging, a quick comparison against other NeoGo peers will reveal the reality (they aren't all "stuck" in the same place).

Any node that isn't progressing should stand out. C# data should never be stale, as noted above, and if the Go node is struggling to progress, it seems like it will be sending pings after all?

The other downside of this approach is that it's only checking connected nodes, we know of many other unconnected peers that could have useful data. I don't imagine there's a way for me to get it without the DoS vector though?

Reply:
There's no way to see when the last `ping` from a node was, right? Just thinking adding a `lastupdated` field might help further differentiate Go nodes that are stuck vs healthy.

Reply:
> There's no way to see when the last ping from a node was, right?

Yes, that's right.

> Just thinking adding a lastupdated field

May be something like this, but only if @roman-khimov think it worths implementing:
```
// HandlePing handles a ping message received from the peer.
func (p *TCPPeer) HandlePing(ping *payload.Ping) error {
	p.lock.Lock()
	defer p.lock.Unlock()
	p.lastBlockIndex = ping.LastBlockIndex
	p.lastUpdated = p.server.chain.BlockHeight()
	return nil
}
```

Reply:
@EdgeDLT, @ixje, we've discussed this issue with Roman and came to the following solution: this change doesn't deserve new fully-qualified RPC handler, but at the same time it may be useful sometimes. Let's extend the existing `getpeers` RPC response with two fields for every connected peer: `useragent` and `lastKnownHeight`. These fields should be marked by `omitempty` JSON tag in order not to marshal them if not specified. Also we've decided not to include `lastUpdated` right now, but if needed then we may extend `getpeers` API one more time to include the *time* of the last peer ping.

Brett, are you interested in the updated version? Could you adjust your implementation?

Reply:
`s/lastKnownHeight/lastknownheight/`, Neo JSON fields are named this way.

Reply:
@AnnaShaleva I've made the changes as requested, tests okay on my end. Please review.